### PR TITLE
Delete users under auth_mode other than db_auth

### LINKED ISSUE #8424: Can not delete user login with oidc
I use `dex` as `oidc provider`. when delete a user from `dex`. user can not login harbor dashboard, but  can still  use `docker` command. I can't delete user in harbor dashboard

image

### PR DESCRIPTION
**Pull request by stonezdj**
The following information should cleanup before delete user:
  Delete project member of this user.
  Delete oidc_user when auth_mode is oidc_auth.
  Fixes #8424

Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: src/controller/user/controller.go [hunk 1]
```diff
@@ -167,6 +173,22 @@ func (c *controller) Count(ctx context.Context, query *q.Query) (int64, error) {
 }
 
 func (c *controller) Delete(ctx context.Context, id int) error {
+	// cleanup project member with the user
+	if err := c.memberMgr.DeleteMemberByUserID(ctx, id); err != nil {
```
Comment (reasonerjt) [inline src/controller/user/controller.go:177]: if the consequent deletion fails how can the membership be restored?

File: src/pkg/oidc/dao/meta.go [hunk 2]
```diff
@@ -42,6 +43,16 @@ func NewMetaDao() MetaDAO {
 
 type metaDAO struct{}
 
+func (md *metaDAO) DeleteByUserId(ctx context.Context, uid int) error {
```
Comment (reasonerjt) [inline src/pkg/oidc/dao/meta.go:46]: Id -> ID

File: src/controller/user/controller.go [hunk 3]
```diff
@@ -167,6 +173,22 @@ func (c *controller) Count(ctx context.Context, query *q.Query) (int64, error) {
...
+		return err
+	}
+
+	// delete oidc metadata under the user
+	a, err := config.AuthMode(ctx)
```
Comment (reasonerjt) [inline src/controller/user/controller.go:182]: When the func in controller is called, he authmode is very likely populated to the ctx already:
https://github.com/goharbor/harbor/blob/60478f4990975ffc878f0b186843e7c0f5f87c52/src/server/middleware/security/security.go#L54

File: src/pkg/oidc/dao/meta.go [hunk 4]
```diff
@@ -29,7 +29,8 @@ type MetaDAO interface {
 	Create(ctx context.Context, oidcUser *models.OIDCUser) (int, error)
 	// GetByUsername get the oidc meta record by the user's username
 	GetByUsername(ctx context.Context, username string) (*models.OIDCUser, error)
-	// Update ...
+	// DeleteByUserID delete the oidc meta by user id
+	DeleteByUserId(ctx context.Context, uid int) error // Update ...
 	Update(ctx context.Context, oidcUser *models.OIDCUser, props ...string) error
```
Comment (reasonerjt) [inline src/pkg/oidc/dao/meta.go:34]: The comment of `Update` is gone.

File: src/pkg/oidc/dao/meta.go (hunk 2 above)
Comment (stonezdj) [reply src/pkg/oidc/dao/meta.go:46]: Updated

File: src/controller/user/controller.go (hunk 3 above)
Comment (stonezdj) [reply src/controller/user/controller.go:182]: Updated

File: src/pkg/oidc/dao/meta.go (hunk 4 above)
Comment (stonezdj) [reply src/pkg/oidc/dao/meta.go:34]: Fixed

File: src/pkg/oidc/dao/meta.go [hunk 5]
```diff
@@ -29,6 +29,8 @@ type MetaDAO interface {
 	Create(ctx context.Context, oidcUser *models.OIDCUser) (int, error)
 	// GetByUsername get the oidc meta record by the user's username
 	GetByUsername(ctx context.Context, username string) (*models.OIDCUser, error)
+	// DeleteByUserID delete the oidc meta by user id
```
Comment (lindhe) [inline src/pkg/oidc/dao/meta.go:32]: typo: meta --> metadata

Comment (stonezdj) [reply src/pkg/oidc/dao/meta.go:32]: updated

File: src/controller/user/controller.go [hunk 6]
```diff
@@ -167,6 +172,16 @@ func (c *controller) Count(ctx context.Context, query *q.Query) (int64, error) {
 }
 
 func (c *controller) Delete(ctx context.Context, id int) error {
+	// cleanup project member with the user
+	if err := c.memberMgr.DeleteMemberByUserID(ctx, id); err != nil {
+		return err
```
Comment (reasonerjt) [inline src/controller/user/controller.go:177]: The error can be confusing b/c it doesn't contain information to explicitly let user know it happens when deleting the membership.

Comment (stonezdj) [reply src/controller/user/controller.go]: Updated

Comment (stonezdj) [reply src/controller/user/controller.go]: updated

**Review by reasonerjt — APPROVED** [review]

**Review by heww — APPROVED** [review]

**Comment by lindhe** [comment]
I have forgotten to thank you for implementing this. It's very useful, thanks a lot! :partying_face:

**Pull request merged**